### DISCUSSION TITLE
Variables substitusion

### DISCUSSION THREAD

**Discussion by mardiros** [mardiros/plaster-yaml#27] [Ideas]
Hey @kpinc 

I've setup discussions on plaster-yaml project in order to have a proper place to dicuss about variable substitution.

To me variables should comes from environment variable, because it is my personal needs, and I will step back.

My personal needs is about handling secrets. I want to remove passwords from configuration file, and,

I want to keep code clean without reading environment variable during the configuration phase (I talk
about all pyramid DI here, the includeme function). 

Also, I run http applications in docker containers, and I am working hard on a migration into a Nomad cluster.
Secrets in Nomad are injected using environment variables.

This is why I my source of parameters is the environment variables.

Furthermore, pull requests at Pylons is hard. From my experience, they have their own view of
their products and its hard to get something merged overthere.

For this reason I am not sure that starts replacing variable in plater is really pragmatic.

Thats said, I also have another project that load configuration from yaml file, the [celery-yaml](https://github.com/mardiros/celery-yaml) library that have the same problem. It configure a Celery application from a yaml file but we have
to read secrets from another mechanism if we want to have that.

**Comment by kpinc** [comment]
On Sun, 20 Oct 2024 04:21:59 -0700
Guillaume Gauvrit ***@***.***> wrote:

> I've setup discussions on plaster-yaml project in order to have a
> proper place to dicuss about variable substitution.

Good idea.

> To me variables should comes from environment variable, because it is
> my personal needs, and I will step back.
> 
> My personal needs is about handling secrets. I want to remove
> passwords from configuration file, and,
> 
> I want to keep code clean without reading environment variable during
> the configuration phase (I talk about all pyramid DI here, the
> includeme function). 

So, you want to keep from having to have your app (pyramid's
includeme function) having special environment reading code, etc.

> Also, I run http applications in docker containers, and I am working
> hard on a migration into a Nomad cluster. Secrets in Nomad are
> injected using environment variables.
> 
> This is why I my source of parameters is the environment variables.

It makes complete sense to me.  It's clear that environment variables
are a required data source.

> Furthermore, pull requests at Pylons is hard. From my experience,
> they have their own view of their products and its hard to get
> something merged overthere.
> 
> For this reason I am not sure that starts replacing variable in
> plater is really pragmatic.

My theory is that most of the Pylons people have moved on or
lost interest.  It seems hard to get anything changed, or even
do minimal maintenance in some cases.  Sad, but the state of things.

That said, I think interests align.  And forking is always an option;
if Pylons isn't changing then keeping up with changes is simple.  ;-)

> Thats said, I also have another project that load configuration from
> yaml file, the [celery-yaml](https://github.com/mardiros/celery-yaml)
> library that have the same problem. It configure a Celery application
> from a yaml file but we have to read secrets from another mechanism
> if we want to have that.

I'm not qualified to comment just now.

I do think that jinja2 is a really good fit.  It's expressive enough,
without allowing arbitrary Python operations.  The Python os.environ
value can be dropped into the global namespace so it's available
to all templates.  And most importantly, it's widely used and
supported throughout the Python ecosystem.  (And ansible uses it,
etc.)

Regards,

Karl ***@***.***>
Free Software:  "You don't pay back, you pay forward."
                 -- Robert A. Heinlein

**Comment by mardiros** [comment]
I use Jinja2 a lot but I think it is too much for the job.

I've start yet another project here: https://github.com/mardiros/envsub

**Comment by kpinc** [comment]
On Fri, 25 Oct 2024 11:50:37 -0700
Guillaume Gauvrit ***@***.***> wrote:

> I use Jinja2 a lot but I think it is too much for the job.

Fair enough.  Although I do like the idea of being able
to use expressions {{myvar + 1}} and so forth, which
pretty much takes you straight to "too much"-ish.

What I like about Jinja2 is that it does put limits
in place, to keep your template authors from being too-stupid,
while still allowing basic expressiveness.

Regards,

Karl ***@***.***>
Free Software:  "You don't pay back, you pay forward."
                 -- Robert A. Heinlein

**Reply by mardiros** [reply]
I think that having computation inside the configuration means that you require something like ansible/puppet
to prepare your configuration in that case.

This is too much work for a substitution, I want it to be fast and simple.
Adding jinja2 is opening a door to configuration generator. Too many much responsability for this library.
